File: native/comparison/fastread-window.c
#define _POSIX_C_SOURCE 200809L

#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

enum { BUFFER_SIZE = 1 << 16 };

typedef struct {
	char *data;
	size_t len;
	size_t cap;
} buffer_t;

static void fail(const char *message) {
	perror(message);
	exit(errno ? errno : 1);
}

static void ensure_capacity(buffer_t *buffer, size_t needed) {
	if (needed <= buffer->cap) return;
	size_t next = buffer->cap == 0 ? BUFFER_SIZE : buffer->cap;
	while (next < needed) {
		next *= 2;
	}
	char *data = realloc(buffer->data, next);
	if (!data) fail("realloc");
	buffer->data = data;
	buffer->cap = next;
}

static void append_bytes(buffer_t *buffer, const char *src, size_t count) {
	ensure_capacity(buffer, buffer->len + count + 1);
	memcpy(buffer->data + buffer->len, src, count);
	buffer->len += count;
	buffer->data[buffer->len] = '\0';
}

static void write_all(int fd, const char *buffer, size_t size) {
	size_t written = 0;
	while (written < size) {
		ssize_t bytes_written = write(fd, buffer + written, size - written);
		if (bytes_written < 0) {
			if (errno == EINTR) continue;
			fail("write");
		}
		written += (size_t)bytes_written;
	}
}

static size_t format_size(size_t bytes, char *out, size_t out_size) {
	const char *units[] = {"B", "KB", "MB", "GB"};
	double value = (double)bytes;
	size_t unit = 0;
	while (value >= 1024.0 && unit < 3) {
		value /= 1024.0;
		unit += 1;
	}
	if (unit == 0) {
		return (size_t)snprintf(out, out_size, "%zu %s", bytes, units[unit]);
	}
	return (size_t)snprintf(out, out_size, "%.1f %s", value, units[unit]);
}

int main(int argc, char **argv) {
	if (argc != 4) {
		fprintf(stderr, "usage: %s <file> <offset> <limit>\n", argv[0]);
		return 1;
	}

	const char *path = argv[1];
	long offset_long = strtol(argv[2], NULL, 10);
	long limit_long = strtol(argv[3], NULL, 10);
	if (offset_long < 1 || limit_long < 1) {
		fprintf(stderr, "offset and limit must be >= 1\n");
		return 1;
	}

	size_t start_line = (size_t)offset_long;
	size_t max_lines = (size_t)limit_long;
	const size_t max_bytes = 256000;

	int fd = open(path, O_RDONLY | O_CLOEXEC);
	if (fd < 0) fail("open");

	char chunk[BUFFER_SIZE];
	buffer_t carry = {0};
	size_t current_line = 1;
	size_t output_lines = 0;
	size_t output_bytes = 0;
	bool hit_line_limit = false;
	bool hit_byte_limit = false;
	bool wrote_output = false;
	size_t first_line_excess = 0;

	for (;;) {
		ssize_t bytes_read = read(fd, chunk, sizeof(chunk));
		if (bytes_read == 0) break;
		if (bytes_read < 0) {
			if (errno == EINTR) continue;
			fail("read");
		}

		append_bytes(&carry, chunk, (size_t)bytes_read);
		size_t line_start = 0;

		for (size_t i = 0; i < carry.len; i += 1) {
			if (carry.data[i] != '\n') continue;
			size_t line_len = i - line_start + 1;
			if (current_line >= start_line) {
				if (output_lines >= max_lines) {
					hit_line_limit = true;
					goto done;
				}
				if (output_bytes + line_len > max_bytes) {
					hit_byte_limit = true;
					if (output_lines == 0) first_line_excess = line_len;
					goto done;
				}
				write_all(STDOUT_FILENO, carry.data + line_start, line_len);
				wrote_output = true;
				output_lines += 1;
				output_bytes += line_len;
			}
			current_line += 1;
			line_start = i + 1;
		}

		if (line_start > 0) {
			size_t remaining = carry.len - line_start;
			memmove(carry.data, carry.data + line_start, remaining);
			carry.len = remaining;
			carry.data[carry.len] = '\0';
		}
	}

	if (carry.len > 0 && current_line >= start_line) {
		if (output_lines >= max_lines) {
			hit_line_limit = true;
			goto done;
		}
		if (output_bytes + carry.len > max_bytes) {
			hit_byte_limit = true;
			if (output_lines == 0) first_line_excess = carry.len;
			goto done;
		}
		write_all(STDOUT_FILENO, carry.data, carry.len);
		wrote_output = true;
		output_lines += 1;
		output_bytes += carry.len;
	}

done:
	close(fd);

	if (hit_line_limit) {
		char message[256];
		size_t end_line = start_line + output_lines - 1;
		size_t next_offset = end_line + 1;
		int len = snprintf(
			message,
			sizeof(message),
			"\n\n[Showing lines %zu-%zu. Use offset=%zu to continue.]",
			start_line,
			end_line,
			next_offset);
		if (len < 0) fail("snprintf");
		write_all(STDOUT_FILENO, message, (size_t)len);
	} else if (hit_byte_limit) {
		char message[256];
		if (output_lines == 0) {
			char size_buf[64];
			char limit_buf[64];
			format_size(first_line_excess, size_buf, sizeof(size_buf));
			format_size(max_bytes, limit_buf, sizeof(limit_buf));
			int len = snprintf(
				message,
				sizeof(message),
				"[Line %zu is %s, exceeds %s limit.]",
				start_line,
				size_buf,
				limit_buf);
			if (len < 0) fail("snprintf");
			write_all(STDOUT_FILENO, message, (size_t)len);
		} else {
			char limit_buf[64];
			size_t end_line = start_line + output_lines - 1;
			size_t next_offset = end_line + 1;
			format_size(output_bytes, limit_buf, sizeof(limit_buf));
			int len = snprintf(
				message,
				sizeof(message),
				"\n\n[Showing lines %zu-%zu (%s limit). Use offset=%zu to continue.]",
				start_line,
				end_line,
				limit_buf,
				next_offset);
			if (len < 0) fail("snprintf");
			write_all(STDOUT_FILENO, message, (size_t)len);
		}
	}

	(void)wrote_output;
	free(carry.data);
	return 0;
}

File: native/fastread-window.zig
const std = @import("std");
const Io = std.Io;
const posix = std.posix;
const system = posix.system;

const max_bytes = 256_000;

fn usage(argv0: []const u8) noreturn {
    std.debug.print("usage: {s} <file> <offset> <limit>\n", .{argv0});
    std.process.exit(1);
}

fn fatal(comptime message: []const u8, err: anyerror) noreturn {
    std.debug.print(message ++ ": {s}\n", .{@errorName(err)});
    std.process.exit(1);
}

fn fatalErrno(comptime message: []const u8, err: std.posix.E) noreturn {
    std.debug.print(message ++ ": {s}\n", .{@tagName(err)});
    std.process.exit(1);
}

fn writeAll(bytes: []const u8) void {
    var written: usize = 0;
    while (written < bytes.len) {
        const rc = system.write(1, bytes[written..].ptr, bytes.len - written);
        switch (posix.errno(rc)) {
            .SUCCESS => {
                const n: usize = @intCast(rc);
                if (n == 0) fatalErrno("write", .IO);
                written += n;
            },
            .INTR => {},
            else => |err| fatalErrno("write", err),
        }
    }
}

fn formatSize(out: []u8, bytes: usize) []const u8 {
    const units = [_][]const u8{ "B", "KB", "MB", "GB" };
    var value: f64 = @floatFromInt(bytes);
    var unit: usize = 0;
    while (value >= 1024.0 and unit < units.len - 1) : (unit += 1) {
        value /= 1024.0;
    }
    if (unit == 0) {
        return std.fmt.bufPrint(out, "{} {s}", .{ bytes, units[unit] }) catch unreachable;
    }
    return std.fmt.bufPrint(out, "{d:.1} {s}", .{ value, units[unit] }) catch unreachable;
}

fn parsePositive(value: []const u8) ?usize {
    const parsed = std.fmt.parseInt(i64, value, 10) catch return null;
    if (parsed < 1) return null;
    return @intCast(parsed);
}

fn emitLineLimit(start_line: usize, output_lines: usize) void {
    var buf: [256]u8 = undefined;
    const end_line = start_line + output_lines - 1;
    const next_offset = end_line + 1;
    const text = std.fmt.bufPrint(&buf, "\n\n[Showing lines {}-{}. Use offset={} to continue.]", .{ start_line, end_line, next_offset }) catch unreachable;
    writeAll(text);
}

fn emitByteLimit(start_line: usize, output_lines: usize, output_bytes: usize, first_line_excess: usize) void {
    var buf: [256]u8 = undefined;
    if (output_lines == 0) {
        var size_buf: [64]u8 = undefined;
        var limit_buf: [64]u8 = undefined;
        const text = std.fmt.bufPrint(&buf, "[Line {} is {s}, exceeds {s} limit.]", .{
            start_line,
            formatSize(&size_buf, first_line_excess),
            formatSize(&limit_buf, max_bytes),
        }) catch unreachable;
        writeAll(text);
    } else {
        var limit_buf: [64]u8 = undefined;
        const end_line = start_line + output_lines - 1;
        const next_offset = end_line + 1;
        const text = std.fmt.bufPrint(&buf, "\n\n[Showing lines {}-{} ({s} limit). Use offset={} to continue.]", .{
            start_line,
            end_line,
            formatSize(&limit_buf, output_bytes),
            next_offset,
        }) catch unreachable;
        writeAll(text);
    }
}

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const arena = init.arena.allocator();
    const args = try init.minimal.args.toSlice(arena);
    if (args.len != 4) usage(args[0]);

    const path = args[1];
    const start_line = parsePositive(args[2]) orelse {
        std.debug.print("offset and limit must be >= 1\n", .{});
        std.process.exit(1);
    };
    const max_lines = parsePositive(args[3]) orelse {
        std.debug.print("offset and limit must be >= 1\n", .{});
        std.process.exit(1);
    };

    const stat = Io.Dir.cwd().statFile(io, path, .{}) catch |err| fatal("stat", err);
    if (stat.size == 0) return;
    if (stat.size > std.math.maxInt(usize)) fatal("stat", error.FileTooBig);
    const file_size: usize = @intCast(stat.size);

    const fd = posix.openat(posix.AT.FDCWD, path, .{ .ACCMODE = .RDONLY, .CLOEXEC = true }, 0) catch |err| fatal("open", err);
    defer _ = system.close(fd);

    const mapped = posix.mmap(null, file_size, .{ .READ = true }, .{ .TYPE = .PRIVATE }, fd, 0) catch |err| fatal("mmap", err);
    defer posix.munmap(mapped);
    const data: []const u8 = mapped;

    var current_line: usize = 1;
    var line_start: usize = 0;
    var output_lines: usize = 0;
    var output_bytes: usize = 0;
    var hit_line_limit = false;
    var hit_byte_limit = false;
    var first_line_excess: usize = 0;

    var i: usize = 0;
    while (i < data.len) : (i += 1) {
        if (data[i] != '\n') continue;
        const line_len = i - line_start + 1;
        if (current_line >= start_line) {
            if (output_lines >= max_lines) {
                hit_line_limit = true;
                break;
            }
            if (output_bytes + line_len > max_bytes) {
                hit_byte_limit = true;
                if (output_lines == 0) first_line_excess = line_len;
                break;
            }
            writeAll(data[line_start .. i + 1]);
            output_lines += 1;
            output_bytes += line_len;
        }
        current_line += 1;
        line_start = i + 1;
    }

    if (!hit_line_limit and !hit_byte_limit and line_start < data.len and current_line >= start_line) {
        const line_len = data.len - line_start;
        if (output_lines >= max_lines) {
            hit_line_limit = true;
        } else if (output_bytes + line_len > max_bytes) {
            hit_byte_limit = true;
            if (output_lines == 0) first_line_excess = line_len;
        } else {
            writeAll(data[line_start..]);
            output_lines += 1;
            output_bytes += line_len;
        }
    }

    if (hit_line_limit) {
        emitLineLimit(start_line, output_lines);
    } else if (hit_byte_limit) {
        emitByteLimit(start_line, output_lines, output_bytes, first_line_excess);
    }
}
